security/rbac/service.go

```diff
@@ -0,0 +1,157 @@
+package rbac
+
+import (
+	"errors"
+	"log"
+
+	"github.com/casbin/casbin/v2"
+	"github.com/casbin/casbin/v2/model"
+	"github.com/casbin/casbin/v2/persist"
+	"gopkg.in/yaml.v2"
+
+	"github.com/gaia-pipeline/gaia"
+	"github.com/gaia-pipeline/gaia/helper/assethelper"
+)
+
+type (
+	// RoleRule represents a Casbin role rule line in the format we expect.
+	RoleRule struct {
+		Namespace string `json:"namespace"`
+		Action    string `json:"action"`
+		Resource  string `json:"resource"`
+		Effect    string `json:"effect"`
+	}
+
+	// Service wraps the Casbin enforcer and performs all actions we require to manage and use RBAC functions.
+	Service interface {
+		EndpointEnforcer
+		AddRole(role string, roleRules []RoleRule) error
+		DeleteRole(role string) error
+		GetAllRoles() []string
+		GetUserAttachedRoles(username string) ([]string, error)
+		GetRoleAttachedUsers(role string) ([]string, error)
+		AttachRole(username string, role string) error
+		DetachRole(username string, role string) error
+	}
+
+	enforcerService struct {
+		adapter         persist.BatchAdapter
+		enforcer        casbin.IEnforcer
+		rbacapiMappings gaia.RBACAPIMappings
+	}
+)
+
+// NewEnforcerSvc creates a new EnforcerService.
+func NewEnforcerSvc(adapter persist.BatchAdapter) (Service, error) {
+	model, err := loadModel()
+	if err != nil {
+		return nil, err
+	}
+
+	enforcer, err := casbin.NewEnforcer(model, adapter)
+	if err != nil {
+		log.Fatal(err)
+	}
+	enforcer.EnableLog(true)
+
+	rbacapiMappings, err := loadAPIMappings()
+	if err != nil {
+		return nil, err
+	}
+
+	return &enforcerService{
+		enforcer:        enforcer,
+		rbacapiMappings: rbacapiMappings,
+	}, nil
+}
+
+func loadModel() (model.Model, error) {
+	modelStr, err := assethelper.LoadRBACModel()
+	if err != nil {
+		return nil, err
```

Comment:
Add logging here as well, and all the rest of the errors. The more messages we get the more we know what went wrong. Pass in the logger as a dependency so you can use it with a method receiver.

Reply:
WIll go through and do this as I tidy up mate

Reply:
I'd prefer to add logging at the caller level. So in this case it would be where we call `NewEnforcerSvc`. The rest of the logging I will add into the handler code, etc.

One thing I will do is wrap the errors though to give more context :+1:

Reply:
That's good as well. High level logging is fine but that logging wraps at least 3 or more different errors. If even one of them is similar we'll have no idea where the error happened. Wrapping is a good solution to that. Love the new error system. 😊

Reply:
Where’s a good old Java stacktrace when you need it! Haha

Reply:
Bwuhahaha, pssss, not so loud.. The other gophers might hear you. 😂😂😂